src/shared/validation.ts

```diff
@@ -164,6 +162,128 @@ export const isStateVersion = t.applyCascade(t.isNumber(), [
+        type: "modifier";
+        number: number;
+        damage: RRDamageType;
+      }
+    | {
+        id: RRDiceTemplatePartID;
+        type: "linkedModifier";
+        damage: RRDamageType;
+        name: "initiative";
+      }
+    | {
+        id: RRDiceTemplatePartID;
+        type: "linkedProficiency";
+        damage: RRDamageType;
+        proficiency: typeof proficiencyValues[number];
+      }
+    | {
+        id: RRDiceTemplatePartID;
+        type: "linkedStat";
+        name: typeof characterStatNames[number];
+        damage: RRDamageType;
+      }
+  )[];
+  rollType: typeof rollType[number];
+}
+
+const __isDiceTemplateRecursive = t.makeValidator({
+  test: (value, state): value is _RRDiceTemplate =>
+    isDiceTemplate(value, state),
+});
+
+const isDiceTemplate = t.isObject({
+  id: isRRID<RRDiceTemplateID>(),
+  name: t.isString(),
+  notes: t.isString(),
+  parts: t.applyCascade(
+    t.isArray(
+      t.isOneOf(
+        withDo({ id: isRRID<RRDiceTemplatePartID>() }, (sharedValidators) => [
+          t.isObject({
+            ...sharedValidators,
+            type: t.isLiteral("dice"),
+            count: t.applyCascade(t.isNumber(), [
+              t.isInteger(),
+              t.isPositive(),
+            ]),
+            faces: t.applyCascade(t.isNumber(), [
+              t.isInteger(),
+              t.isPositive(),
+            ]),
+            negated: t.isBoolean(),
+            damage: isDamageType,
+            modified: t.isEnum(multipleRollValues),
+          }),
+          t.isObject({
+            ...sharedValidators,
+            type: t.isLiteral("template"),
+            template: __isDiceTemplateRecursive,
+          }),
+          t.isObject({
+            ...sharedValidators,
+            type: t.isLiteral("modifier"),
+            number: t.applyCascade(t.isNumber(), [t.isInteger()]),
+            damage: isDamageType,
+          }),
+          t.isObject({
+            ...sharedValidators,
+            type: t.isLiteral("linkedModifier"),
+            name: t.isLiteral("initiative"),
+            damage: isDamageType,
+          }),
+          t.isObject({
+            ...sharedValidators,
+            type: t.isLiteral("linkedProficiency"),
+            damage: isDamageType,
+            proficiency: t.isEnum(proficiencyValues),
+          }),
+          t.isObject({
+            ...sharedValidators,
+            type: t.isLiteral("linkedStat"),
+            name: t.isEnum(characterStatNames),
+            damage: isDamageType,
+          }),
+        ])
+      )
+    ),
+    [t.hasUniqueItems({ map: (part) => part.id })]
+  ),
+  rollType: t.isEnum(rollType),
+});
+
+export type RRDiceTemplate = t.InferType<typeof isDiceTemplate>;
+
+assert<IsExact<RRDiceTemplate, _RRDiceTemplate>>(true);
+
+const isDiceTemplateCategory = t.isObject({
+  id: isRRID<RRDiceTemplateCategoryID>(),
+  icon: t.isEnum(categoryIcons),
+  categoryName: t.isString(),
+  templates: t.isArray(isDiceTemplate),
```

Comment:
After migrating my state, I have some dice templates with the same id now. This might totally be because I my state was in a weird state that the migration does not account for and it might not happen with our production state. Could we adjust the validation here to check for duplicate ids and test it on our production state?

```ts
  templates: t.applyCascade(t.isArray(isDiceTemplate), [
    t.hasUniqueItems({ map: (diceTemplate) => diceTemplate.id }),
  ]),
```